.github/workflows/ci.yml

```
@@ -0,0 +1,85 @@
+name: CI
+
+on:
+  push:
+    branches:
+      - dev
+      - main
+    tags:
+  pull_request:
+  release:
+
+jobs:
+  check-copyright:
+    name: Copyright
+    runs-on: ubuntu-latest
+    steps:
+      - name: Checkout
+        uses: actions/checkout@v2
+        with:
+          fetch-depth: 0
+      - name: License header
+        run: .scripts/check-copyright-notice.sh
+      - name: Authors
+        run: .scripts/check-notice-authors.sh origin/main
+
+  vet:
+    name: Vetting
+    runs-on: ubuntu-latest
+    steps:
+      - name: Checkout
+        uses: actions/checkout@v2
+      - name: Check vanity import
+        run: .scripts/check-vanity-imports.sh $GITHUB_WORKSPACE
+      - name: Lint
+        uses: golangci/golangci-lint-action@v2
+        with:
+          version: v1.40.1
+
+  tests:
+    name: Tests
+    runs-on: ubuntu-latest
+    steps:
+      - name: Set up Golang
+        uses: actions/setup-go@v2
+        with:
+          go-version: ^1.17
+
+      - name: Checkout
+        uses: actions/checkout@v2
+
+      - uses: actions/cache@v2
+        with:
+          path: |
+            ~/.cache/go-build
+            ~/go/pkg/mod
+          key: ${{ runner.os }}-go-${{ hashFiles('**/go.sum') }}
+          restore-keys: |
+            ${{ runner.os }}-go-
+
+      - name: Unit Tests
+        run: go test -timeout 60s ./...
+
+  race-tests:
+    name: Race Tests
+    runs-on: ubuntu-latest
+    steps:
+      - name: Set up Golang
+        uses: actions/setup-go@v2
+        with:
+          go-version: ^1.17
+
+      - name: Checkout
+        uses: actions/checkout@v2
+
+      - uses: actions/cache@v2
+        with:
+          path: |
+            ~/.cache/go-build
+            ~/go/pkg/mod
+          key: ${{ runner.os }}-go-${{ hashFiles('**/go.sum') }}-race
```

I'm wondering if we can actually share the cache between race and normal tests. The go build should be exactly the same.
```suggestion
          key: ${{ runner.os }}-go-${{ hashFiles('**/go.sum') }}
```

---

Then which version should be used after the tests are run?  
I dont know how GitHub handles merge conflicts in the cache.

---

Caching works as follows: You specify a key and a path. Then, in your pipeline you can load previously stored data from the cache, and you can store the current data in the cache under the specified key. 

The reason why we used a separate key for race tests is that we were running the standard tests and the race tests in parallel. So if we want to cache the test runs for both cases, we need store them under a different key. I would be surprised if the caching tool supports any kind of cache merging.

---

I also don't think that "cache merging" works. I guess the slower job (here: race test) would just overwrite the cache. But the cache content will probably be the exact same. So both tests would just load the same cache data. It would save a duplicate cache entry.

But I don't mind keeping separate caches. Was just an idea to save a duplicate cache.

---

I will try it, lets see what happens.